types/scratch-vm/index.d.ts

```diff
@@ -0,0 +1,199 @@
+    docsURI?: string;
+
+    /** The blocks provided by this extension and the separators. */
+    blocks: Array<ExtensionBlockMetadata | string>;
+
+    /** Map of menu name to metadata for each of this extension's menus. */
+    menus?: Record<string, ExtensionMenuMetadata>;
+}
+
+/** All the metadata needed to register an extension block. */
+interface ExtensionBlockMetadata {
+    /** A unique alphanumeric identifier for this block. No special characters allowed. */
+    opcode: string;
+
+    /** The name of the function implementing this block. Can be shared by other blocks/opcodes. */
+    func?: string;
+
+    /** The type of block (command, reporter, etc.) being described. */
+    blockType: ValueOf<Scratch.extensions.BlockType>;
+
+    /** The text on the block, with [PLACEHOLDERS] for arguments. */
+    text: string;
+
+    /** Whether this block should not appear in the block palette. */
+    hideFromPalette?: boolean;
+
+    /** Whether the block ends a stack - no blocks can be connected after it. */
+    isTerminal?: boolean;
+
+    /** Whether this block is a reporter but should not allow a monitor. */
+    disableMonitor?: boolean;
+
+    /** If this block is a reporter, this is the scope/context for its value. */
+    reporterScope?: ReporterScope;
+
+    /** Whether a hat block is edge-activated. */
+    isEdgeActivated?: boolean;
+
+    /** Whether a hat/event block should restart existing threads. */
+    shouldRestartExistingThreads?: boolean;
+
+    /** For flow control blocks, the number of branches/substacks for this block. */
+    branchCount?: number;
+
+    /** Map of argument placeholder to metadata about each arg. */
+    arguments?: Record<string, ExtensionArgumentMetadata>;
+}
+
+/** All the metadata needed to register an argument for an extension block. */
+interface ExtensionArgumentMetadata {
+    /** The type of the argument (number, string, etc.) */
+    type: ValueOf<Scratch.extensions.ArgumentType>;
+
+    /** The default value of this argument. */
+    defaultValue?: any;
+
+    /** The name of the menu to use for this argument, if any. */
+    menu?: string;
+}
+
+/** All the metadata needed to register an extension drop-down menu. */
+type ExtensionMenuMetadata = ExtensionDynamicMenu | ExtensionMenuItems;
+
+/** The string name of a function which returns menu items. */
+type ExtensionDynamicMenu = string;
+
+/** Items in an extension menu. */
+type ExtensionMenuItems = Array<ExtensionMenuItemSimple | ExtensionMenuItemComplex>;
+
+/** A menu item for which the label and value are identical strings. */
+type ExtensionMenuItemSimple = string;
+
+/** A menu item for which the label and value can differ. */
+interface ExtensionMenuItemComplex {
+    /** The value of the block argument when this menu item is selected. */
+    value: any;
+
+    /** The human-readable label of this menu item in the menu. */
+    text: string;
+}
+
+interface ScratchExtension {
```

Comment:
IMO this lacks the constructor that accepts the `runtime`:
https://github.com/LLK/scratch-vm/blob/develop/docs/extensions.md#defining-an-extension
e.g.:
```ts
new (runtime: ScratchRuntime): ScratchExtension;
```
am I correct wit this?

Reply:
You are correct with this being a class but not correct about taking a runtime parameter.

Reply:
Fixed!